File: docs/interfaces/ios/DigitalHumanInterface.swift
import Foundation

// ==================== 统一数字人接口 - iOS 实现 ====================
// 实现 docs/interfaces/UNIFIED_INTERFACE.md 中定义的接口
//
// @since 2026-04-14
// @status 🟡 进行中

// MARK: - DigitalHuman 主接口

public protocol DigitalHuman: AnyObject {
    // 生命周期
    func start() async -> Bool  // 启动数字人
    func stop() async -> Bool   // 停止数字人
    var isReady: Bool { get }    // 是否就绪
    
    // 音频驱动（唇形同步）
    func sendAudio(_ audioData: Data)   // 发送音频数据驱动口型
    func sendText(_ text: String)        // 发送文字（备用）
    
    // 语音交互
    func speak(_ text: String) async -> Bool  // 数字人说话
    func stopSpeaking()                        // 停止说话
    
    // 配置
    func setStyle(_ style: HumanStyle)
    func setEmotion(_ emotion: HumanEmotion)
}

// MARK: - 枚举定义

public enum HumanStyle {
    case cartoon      // 卡通风格
    case realistic    // 真人风格
    case style2D      // 2D 风格
}

public enum HumanEmotion {
    case neutral      // 中性
    case happy        // 开心
    case serious      // 严肃
    case thinking     // 思考
}

// MARK: - DigitalHumanListener 回调接口

public protocol DigitalHumanListener: AnyObject {
    func onReady()                                    // 数字人准备就绪
    func onError(_ error: Error)                      // 发生错误
    func onSpeakingStarted()                          // 开始说话
    func onSpeakingFinished()                         // 说话结束
    func onAudioLevel(_ level: Float)                // 音频级别（用于UI可视化）
    func onEmotionChanged(_ emotion: HumanEmotion)   // 情绪变化
}

// MARK: - AudioPipeline 接口

public protocol AudioPipeline: AnyObject {
    func start() async -> Bool
    func stop()
    
    // 音频输入
    func sendAudioInput(_ audioData: Data)
    
    // 配置
    func setASRProvider(_ provider: ASRProvider)
    func setTTSProvider(_ provider: TTSProvider)
    func setLLMProvider(_ provider: LLMProvider)
    
    // 回调
    func setListener(_ listener: AudioPipelineListener)
}

// MARK: - AudioPipelineListener 回调接口

public protocol AudioPipelineListener: AnyObject {
    func onAudioInputRecognized(_ text: String)    // ASR 识别结果
    func onLLMResponse(_ text: String)            // LLM 回复
    func onTTSGenerated(_ audioData: Data)        // TTS 音频
    func onError(_ error: Error)
}

// MARK: - Provider 枚举

public enum ASRProvider {
    case volcengine   // 火山引擎
    case other
}

public enum TTSProvider {
    case volcengine   // 火山引擎
    case other
}

public enum LLMProvider {
    case deepseek     // DeepSeek
    case other
}

// MARK: - 工厂函数

/// 创建数字人实例
/// - Parameters:
///   - config: 数字人配置
///   - listener: 事件回调
/// - Returns: DigitalHuman 实例
public func createDigitalHuman(
    config: DigitalHumanConfig,
    listener: DigitalHumanListener
) -> DigitalHuman {
    // TODO: 实现 Live2D 渲染 + 火山引擎 ASR/TTS 集成
    fatalError("等待 Phase 3 实现")
}

/// 创建音频管道实例
/// - Parameters:
///   - config: 管道配置
///   - listener: 事件回调
/// - Returns: AudioPipeline 实例
public func createAudioPipeline(
    config: AudioPipelineConfig,
    listener: AudioPipelineListener
) -> AudioPipeline {
    // TODO: 实现 WebRTC + 火山引擎 ASR + DeepSeek + TTS 管道
    fatalError("等待 Phase 3 实现")
}

// MARK: - 配置结构体

public struct DigitalHumanConfig {
    public let modelPath: String        // Live2D 模型路径
    public let serverUrl: String        // 后端服务器 URL
    public let style: HumanStyle
    public let emotion: HumanEmotion
    
    public init(
        modelPath: String,
        serverUrl: String,
        style: HumanStyle = .cartoon,
        emotion: HumanEmotion = .neutral
    ) {
        self.modelPath = modelPath
        self.serverUrl = serverUrl
        self.style = style
        self.emotion = emotion
    }
}

public struct AudioPipelineConfig {
    public let asrProvider: ASRProvider
    public let ttsProvider: TTSProvider
    public let llmProvider: LLMProvider
    public let apiKeys: [String: String]  // 各 provider 的 API keys
    
    public init(
        asrProvider: ASRProvider = .volcengine,
        ttsProvider: TTSProvider = .volcengine,
        llmProvider: LLMProvider = .deepseek,
        apiKeys: [String: String] = [:]
    ) {
        self.asrProvider = asrProvider
        self.ttsProvider = ttsProvider
        self.llmProvider = llmProvider
        self.apiKeys = apiKeys
    }
}

File: docs/interfaces/android/DigitalHumanInterface.kt
package com.aiinterview.digitalhuman

/**
 * 统一数字人接口 - Android 实现
 * 
 * 实现 docs/interfaces/UNIFIED_INTERFACE.md 中定义的接口
 * 
 * @since 2026-04-14
 * @status 🟡 进行中
 */

interface DigitalHuman {
    // 生命周期
    suspend fun start(): Boolean  // 启动数字人
    suspend fun stop(): Boolean   // 停止数字人
    fun isReady(): Boolean        // 是否就绪
    
    // 音频驱动（唇形同步）
    fun sendAudio(audioData: ByteArray)  // 发送音频数据驱动口型
    fun sendText(text: String)           // 发送文字（备用）
    
    // 语音交互
    suspend fun speak(text: String): Boolean  // 数字人说话
    fun stopSpeaking(): Unit                   // 停止说话
    
    // 配置
    fun setStyle(style: HumanStyle): Unit
    fun setEmotion(emotion: HumanEmotion): Unit
}

enum class HumanStyle {
    CARTOON,     // 卡通风格
    REALISTIC,   // 真人风格
    STYLE_2D     // 2D 风格
}

enum class HumanEmotion {
    NEUTRAL,     // 中性
    HAPPY,       // 开心
    SERIOUS,     // 严肃
    THINKING     // 思考
}

interface DigitalHumanListener {
    fun onReady()                                    // 数字人准备就绪
    fun onError(error: Throwable)                    // 发生错误
    fun onSpeakingStarted()                          // 开始说话
    fun onSpeakingFinished()                         // 说话结束
    fun onAudioLevel(level: Float)                   // 音频级别（用于UI可视化）
    fun onEmotionChanged(emotion: HumanEmotion)     // 情绪变化
}

// ==================== AudioPipeline 接口 ====================

interface AudioPipeline {
    suspend fun start(): Boolean
    fun stop()
    
    // 音频输入
    fun sendAudioInput(audioData: ByteArray)
    
    // 配置
    fun setASRProvider(provider: ASRProvider)
    fun setTTSProvider(provider: TTSProvider)
    fun setLLMProvider(provider: LLMProvider)
    
    // 回调
    fun setListener(listener: AudioPipelineListener)
}

interface AudioPipelineListener {
    fun onAudioInputRecognized(text: String)           // ASR 识别结果
    fun onLLMResponse(text: String)                   // LLM 回复
    fun onTTSGenerated(audioData: ByteArray)          // TTS 音频
    fun onError(error: Throwable)
}

enum class ASRProvider {
    VOLCENGINE,   // 火山引擎
    OTHER
}

enum class TTSProvider {
    VOLCENGINE,   // 火山引擎
    OTHER
}

enum class LLMProvider {
    DEEPSEEK,     // DeepSeek
    OTHER
}

// ==================== 工厂函数 ====================

/**
 * 创建数字人实例
 * 
 * @param config 数字人配置
 * @param listener 事件回调
 * @return DigitalHuman 实例
 */
fun createDigitalHuman(
    config: DigitalHumanConfig,
    listener: DigitalHumanListener
): DigitalHuman {
    // TODO: 实现 Live2D 渲染 + 火山引擎 ASR/TTS 集成
    TODO("等待 Phase 2 实现")
}

data class DigitalHumanConfig(
    val modelPath: String,        // Live2D 模型路径
    val serverUrl: String,        // 后端服务器 URL
    val style: HumanStyle = HumanStyle.CARTOON,
    val emotion: HumanEmotion = HumanEmotion.NEUTRAL
)

/**
 * 创建音频管道实例
 * 
 * @param config 管道配置
 * @param listener 事件回调
 * @return AudioPipeline 实例
 */
fun createAudioPipeline(
    config: AudioPipelineConfig,
    listener: AudioPipelineListener
): AudioPipeline {
    // TODO: 实现 WebRTC + 火山引擎 ASR + DeepSeek + TTS 管道
    TODO("等待 Phase 2 实现")
}

data class AudioPipelineConfig(
    val asrProvider: ASRProvider = ASRProvider.VOLCENGINE,
    val ttsProvider: TTSProvider = TTSProvider.VOLCENGINE,
    val llmProvider: LLMProvider = LLMProvider.DEEPSEEK,
    val apiKeys: Map<String, String>  // 各 provider 的 API keys
)